include/aspect/particle/property/crystal_preferred_orientation.h

```diff
@@ -0,0 +1,387 @@
+ */
+
+#ifndef _aspect_particle_property_cpo_h
+#define _aspect_particle_property_cpo_h
+
+#include <aspect/particle/property/interface.h>
+#include <aspect/simulator_access.h>
+#include <array>
+
+DEAL_II_DISABLE_EXTRA_DIAGNOSTICS
+#include <boost/random.hpp>
+DEAL_II_ENABLE_EXTRA_DIAGNOSTICS
+
+namespace aspect
+{
+  namespace Particle
+  {
+    namespace Property
+    {
+      enum class DeformationType
+      {
+        passive
+      };
+
+      enum class DeformationTypeSelector
+      {
+        passive
+      };
+
+      enum class AdvectionMethod
+      {
+        forward_euler, backward_euler, crank_nicolson
+      };
+
+      enum class CPODerivativeAlgorithm
+      {
+        spin_tensor, drex_2004
+      };
+
+      /**
+       * The plugin manages and computes the evolution of Lattice/Crystal Preferred Orientations (LPO/CPO)
+       * on particles. Each ASPECT particle represents many grains. Each grain is assigned a size and a orientation
+       * matrix. This allows tracking the LPO evolution with kinematic polycrystal CPO evolution models such
+       * as D-Rex (Kaminski and Ribe, 2001; ́Kaminski et al., 2004).
+       *
+       * This plugin stores M minerals and for each mineral it stores N grains. The total amount of memory stored is 2
+       * doubles per mineral plus 10 doubles per grain, resulting in a total memory of M * (2 + N * 10) . The layout of
+       * the data stored is the following (note that for this plugin the following dims are always 3):
+       * 1. M minerals times
+       *    1.1  Mineral deformation type   -> 1 double, at location
+       *                                      => 0 + mineral_i * (n_grains * 10 + 2)
+       *    2.1. Mineral volume fraction    -> 1 double, at location
+       *                                      => 1 + mineral_i * (n_grains * 10 + 2)
+       *    2.2. N grains times:
+       *         2.1. volume fraction grain -> 1 double, at location:
+       *                                      => 2 + grain_i * 10 + mineral_i * (n_grains * 10 + 2)
+       *         2.2. rotation_matrix grain -> 9 (Tensor<2,dim>::n_independent_components) doubles, starts at:
+       *                                      => 3 + grain_i * 10 + mineral_i * (n_grains * 10 + 2)
+       *
+       * Last used data entry is n_minerals * (n_grains * 10 + 2). In case the Crank-Nicolson
+       * advection scheme is used, the derivative data is stored after the main data and each grain
+       * stores twice the amount of data. The last data entry in this case is:
+       * n_minerals * (n_grains * 10 + 2) + n_minerals * (n_grains * 10).
+       *
+       * We store the same number of grains for all minerals (e.g. olivine and enstatite
+       * grains), although their volume fractions may not be the same. This is because we need a minimum number
+       * of grains per tracer to perform reliable statistics on it. This minimum should be the same for both
+       * olivine and enstatite.
```

Comment:
mineral specific, such it be?

Reply:
I will make this into `This minimum should be the same for all minerals.`